internal/k8s/reconciler/state/gateway.go

```diff
@@ -0,0 +1,104 @@
+package state
+
+import (
+	"errors"
+
+	"github.com/hashicorp/consul-api-gateway/internal/core"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/common"
+	"github.com/hashicorp/consul-api-gateway/internal/k8s/reconciler/status"
+	gw "sigs.k8s.io/gateway-api/apis/v1alpha2"
+)
+
+// GatewayState holds ephemeral state for gateways
+type GatewayState struct {
+	ConsulNamespace string
+	Status          status.GatewayStatus
+	PodReady        bool
+	ServiceReady    bool
+	Generation      int64
+	Addresses       []string
+
+	Listeners []*ListenerState
+}
+
+func InitialGatewayState(g *gw.Gateway) *GatewayState {
+	state := &GatewayState{
+		Generation: g.GetGeneration(),
+	}
+	for _, listener := range g.Spec.Listeners {
+		state.Listeners = append(state.Listeners, &ListenerState{
+			Name:     listener.Name,
+			Protocol: listener.Protocol,
+			Routes:   make(map[string]core.ResolvedRoute),
+		})
+	}
+	return state
+}
+
+func (g *GatewayState) GetStatus(gateway *gw.Gateway) gw.GatewayStatus {
```

Comment:
Don't forget to add a few tests in future PRs. This seems like a critical bit to ensure correctness of.

Reply:
👍 the testing code for this is going to get a refactor along with moving status setting out of `gateway.go` and into a set of dedicated validators.